File: Mappy/Mappy.py
__author__ = 'Skully'
__version__ = '2.0'

import clr
import sys
import math

clr.AddReferenceByPartialName("UnityEngine")
clr.AddReferenceByPartialName("Pluton")
clr.AddReference("Assembly-CSharp")
import World as globalWorld
import UnityEngine
import Pluton
import System
from System import Uri

class Mappy:
    def ConfigurationFile(self):
        if not Plugin.IniExists("ConfigurationFile"):
            ini = Plugin.CreateIni("ConfigurationFile")
            ini.AddSetting("Settings", "enabled", "1")
            ini.AddSetting("Settings", "SendChat", "1")
            ini.AddSetting("Settings", "Timer", "60000")
            ini.AddSetting("Settings", "url", "http://www.example.com/mappy/")
            ini.Save()
        return Plugin.GetIni("ConfigurationFile")

    def On_PluginInit(self):
        DataStore.Flush("Mappy")
        ini = self.ConfigurationFile()
        if ini.GetSetting("Settings", "enabled") == "1":
            ServerConsoleCommands.Register("mappy").setCallback("ReceiveCommand")
            link = ini.GetSetting("Settings", "url")
            if ini.GetSetting("Settings", "SendChat") == "1":
                DataStore.Add("Mappy", "SendChat", 1)
                DataStore.Add("Mappy", "LinkChat", link + "chat.php")
            DataStore.Add("Mappy", "Link", link + "server.php")
            DataStore.Add("Mappy", "LinkSize", link + "size.php")
            mseconds = ini.GetSetting("Settings", "Timer")
            msec = int(mseconds)
            Plugin.CreateTimer("SendSizeOnce", 15000).Start()
            Plugin.CreateTimer("Send", msec).Start()

    def ReceiveCommand(self, args):
        if len(args) == 0:
            return
        if args[0] == "airdrop":
            x = args[1]
            z = args[2]
            World.AirDropAt(float(x), 0, float(z))
        elif args[0] == "kick":
            steamid = args[1]
            palyer = Pluton.Player.Find(steamid)
            player.Kick("Mappy control panel kick")
        elif args[0] == "message":
            steamid = args[1]
            palyer = Pluton.Player.Find(steamid)
            message = self.MessageMake(args)
            player.Message(message)
        elif args[0] == "teleport":
            x = args[1]
            z = args[2]
            steamid = args[3]
            palyer = Pluton.Player.Find(steamid)
            player.Teleport(float(x), World.GetGround(float(x), float(z)) + 1, float(z))
        elif args[0] == "give":
            steamid = args[1]
            palyer = Pluton.Player.Find(steamid)
            count = args[2]
            # GIVE CODE USED FROM "Give" balu92's PLUGIN
            item = Pluton.InvItem.GetItemID(args[3])
            player.Inventory.Add(item, int(count))
        elif args[0] == "animal":
            x = args[1]
            z = args[2]
            animalname = args[3]
            World.SpawnAnimal(animalname, float(x), float(z))
        elif args[0] == "broadcast":
            message = self.MessageMakeBroadcast(args)
            Server.Broadcast(message)

    def MessageMake(self, args):
        message = str.Join(" ", args)
        message = message.replace(args[0] + " " + args[1] + " ", "")
        return message

    def MessageMakeBroadcast(self, args):
        message = str.Join(" ", args)
        message = message.replace(args[0] + " ", "")
        return message

    def SendSizeOnceCallback(self, timer):
        timer.Kill()
        DataStore.Add("Mappy", "SizeSent", 1)
        link = DataStore.Get("Mappy", "LinkSize")
        WorldSize = "&worldsize=" + str(globalWorld.Size)
        Plugin.POST(link, WorldSize)

    def SendCallback(self, timer):
        ServersTime = str(World.Time)
        post = "&time=" + ServersTime + "&sleepers=" + str(Server.SleepingPlayers.Count)
        if DataStore.Get("Mappy", "SendChat") == 1:
            post = post + "&showchat=true"
        post = post + "&players=::"
        i = 1
        for player in Server.ActivePlayers:
            if player.Location:
                Name = self.FormatName(player.Name)
                if len(Name) < 3:
                    Name = "Player - " + str(i)
                    i = i+1
                coords = str(player.Location)
                stripit = str.strip(coords, "()")
                sripped = str.strip(stripit, ",")
                splitted = sripped.split(" ")
                coordx = splitted[0]
                coordz = splitted[2]
                steamid = player.SteamID
                post = post + ";" + Name + ":" + coordx + ":" + coordz + ":" + steamid
        link = DataStore.Get("Mappy", "Link")
        Plugin.POST(link, post)

    def On_Chat(self, Chat):
        if DataStore.Get("Mappy", "SendChat") == 1:
            Message = self.FormatChatLine(Chat.OriginalText)
            Sender = self.FormatName(Chat.User.Name)
            if len(Sender) < 3:
                Sender = "Player"
            link = DataStore.Get("Mappy", "LinkChat")
            post = "&chat=" + Sender + ": " + Message
            Plugin.POST(link, post)

    def FormatName(self, Name):
        Name = Uri.EscapeDataString(Name)
        return str(Name)

    def FormatChatLine(self, Line):
        Line = Uri.EscapeDataString(Line)
        return str(Line)

File: Mappy.cs
using System;
using UnityEngine;
using Pluton.Core;
using Pluton.Rust;
using Pluton.Rust.Events;
using Pluton.Rust.Objects;
using Pluton.Rust.PluginLoaders;

namespace Mappy
{
    public class Mappy : CSharpPlugin
    {
        bool chatEnabled = false;
        string Link = string.Empty;
        string LinkSize = string.Empty;
        string LinkChat = string.Empty;

        public void On_PluginInit()
        {
            Author = "SkullyDev";
            Version = "1.0";
            About = "";

            if (!Server.Instance.Loaded) return;
            LoadPlugin();
        }

        public void On_ServerInit()
        {
            LoadPlugin();
        }

        private void LoadPlugin()
        {
            IniParser ini = ConfigurationFile();
            if (ini.GetSetting("Settings", "enabled") == "1")
            {
                string link = ini.GetSetting("Settings", "url");
                ServerConsoleCommands.Register("mappy").setCallback(GetCommand);
                if (ini.GetSetting("Settings", "SendChat") == "1")
                {
                    chatEnabled = true;
                    LinkChat = link + "chat.php";
                }
                Link = link + "server.php";
                LinkSize = link + "size.php";
                int mseconds = ToInt(ini.GetSetting("Settings", "Timer"));
                Plugin.CreateTimer("MappySend", mseconds).Start();
                string WorldSize = string.Format("&worldsize={0}", global::World.Size.ToString());
                Plugin.POST(LinkSize, WorldSize);
            }
        }

        public void GetCommand(string[] args)
        {
            if (args[0] != "")
            {
                if (args[0] == "airdrop")
                {
                    if (args.Length >= 3)
                    {
                        string x = args[1];
                        string z = args[2];
                        World.AirDropAt(float.Parse(x), 0, float.Parse(z));
                    }
                }

                else if (args[0] == "kick")
                {
                    if (args.Length >= 2)
                    {
                        ulong sid = ulong.Parse(args[1]);
                        if (Server.Instance.Players.ContainsKey(sid))
                        {
                            Player player = Server.Instance.Players[sid];
                            string message = string.Join(" ", args);
                            message = string.Format("{0}", message.Replace(args[0] + " " + args[1] + " ", ""));
                            player.Kick(message);
                        }
                    }
                }

                else if (args[0] == "ban")
                {
                    if (args.Length >= 2)
                    {
                        ulong sid = ulong.Parse(args[1]);
                        if (Server.Instance.Players.ContainsKey(sid))
                        {
                            Player player = Server.Instance.Players[sid];
                            string message = string.Join(" ", args);
                            message = string.Format("{0}", message.Replace(args[0] + " " + args[1] + " ", ""));
                            player.Ban(message);
                        }
                    }
                }

                else if (args[0] == "message")
                {
                    if (args.Length >= 3)
                    {
                        ulong sid = ulong.Parse(args[1]);
                        if (Server.Instance.Players.ContainsKey(sid))
                        {
                            Player player = Server.Instance.Players[sid];
                            string message = string.Join(" ", args);
                            message = string.Format("{0}", message.Replace(args[0] + " " + args[1] + " ", ""));
                            player.Message(message);
                        }
                    }
                }

                else if (args[0] == "give")
                {
                    if (args.Length >= 4)
                    {
                        ulong sid = ulong.Parse(args[1]);
                        if (Server.Instance.Players.ContainsKey(sid))
                        {
                            Player player = Server.Instance.Players[sid];
                            int count = ToInt(args[2]);
                            int item = 0;
                            item = InvItem.GetItemID(args[3]);
                            if (item == 0) return;
                            player.Inventory.Add(item, count);
                        }
                    }
                }

                else if (args[0] == "teleport")
                {
                    if (args.Length >= 4)
                    {
                        ulong sid = ulong.Parse(args[1]);
                        if (Server.Instance.Players.ContainsKey(sid))
                        {
                            Player player = Server.Instance.Players[sid];
                            string x = args[2];
                            string z = args[3];
                            player.Teleport(float.Parse(x), World.GetGround(float.Parse(x), float.Parse(z)), float.Parse(z));
                        }
                    }
                }

                else if (args[0] == "animal")
                {
                    if (args.Length >= 4)
                    {
                        float x = float.Parse(args[1]);
                        float z = float.Parse(args[2]);
                        string animalname = args[3];
                        GameManager.server.CreateEntity("autospawn/animals/" + animalname, new Vector3(x, World.GetGround(x, z), z)).Spawn(true);
                    }
                }

                else if (args[0] == "broadcast")
                {
                    if (args.Length >= 2)
                    {
                        string message = string.Join(" ", args);
                        message = string.Format("{0}", message.Replace(args[0] + " ", ""));
                        Server.Instance.Broadcast(message);
                    }
                }
            }
            else
            {
                Pluton.Core.Logger.Log("MAPPY PLUGIN & MAP CODE MADE BY Skully (SkullyDev)");
                Pluton.Core.Logger.Log("Proud member of Pluton-Team.ORG");
            }
        }

        private IniParser ConfigurationFile()
        {
            if (!Plugin.IniExists("ConfigurationFile"))
            {
                IniParser ini = Plugin.CreateIni("ConfigurationFile");
                ini.AddSetting("Settings", "enabled", "1");
                ini.AddSetting("Settings", "SendChat", "1");
                ini.AddSetting("Settings", "Timer", "5000");
                ini.AddSetting("Settings", "url", "http://www.example.com/mappy/");
                ini.Save();
            }
            return Plugin.GetIni("ConfigurationFile");
        }

        public void MappySendCallback(TimedEvent timer)
        {
            string post = string.Format("&time={0}&sleepers={1}", TOD_Sky.Instance.Cycle.Hour.ToString(), Server.Instance.SleepingPlayers.Count.ToString());
            if (chatEnabled) post = string.Format("{0}&showchat=true", post);
            post = string.Format("{0}&players=::", post);
            foreach (Player player in Server.Instance.ActivePlayers) post = string.Format("{0};{1}:{2}:{3}:{4}", post, Uri.EscapeDataString(player.Name), player.X, player.Z, player.SteamID);
            try { Plugin.POST(Link, post); } catch { }
        }

        public void On_Chat(ChatEvent Chat)
        {
            if (chatEnabled)
            {
                Player player = Chat.User;
                string post = string.Format("&chat={0}: {1}", Uri.EscapeDataString(player.Name), Uri.EscapeDataString(Chat.OriginalText));
                try { Plugin.POST(LinkChat, post); } catch { }
            }
        }

        private int ToInt(string obj)
        {
            int i;
            int.TryParse(obj, out i);
            return i;
        }
    }
}